File: src/kernel/pes/PEMux.cc
/*
 * Copyright (C) 2016-2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/log/Kernel.h>

#include <thread/ThreadManager.h>

#include "pes/PEMux.h"
#include "pes/VPE.h"
#include "TCU.h"
#include "Platform.h"

namespace kernel {

PEMux::PEMux(peid_t pe)
    : _pe(new PEObject(pe, EP_COUNT - m3::TCU::FIRST_USER_EP)),
      _caps(VPE::INVALID_ID),
      _vpes(),
      _mem_base(),
      _eps(),
      _upcqueue(pe, m3::TCU::PEXUP_REP) {
    for(epid_t ep = 0; ep < m3::TCU::FIRST_USER_EP; ++ep)
        _eps.set(ep);

#if defined(__gem5__)
    if(Platform::pe(pe).supports_pemux()) {
        // configure send EP
        TCU::config_remote_ep(0, pe, m3::TCU::KPEX_SEP, [this](m3::TCU::reg_t *ep_regs) {
            TCU::config_send(ep_regs, m3::KIF::PEMUX_VPE_ID, m3::ptr_to_label(this),
                             Platform::kernel_pe(), TCU::PEX_REP, KPEX_RBUF_ORDER, 1);
        });

        // configure receive EP
        uintptr_t rbuf = Platform::rbuf_pemux(pe);
        TCU::config_remote_ep(0, pe, m3::TCU::KPEX_REP, [rbuf](m3::TCU::reg_t *ep_regs) {
            TCU::config_recv(ep_regs, m3::KIF::PEMUX_VPE_ID, rbuf,
                             KPEX_RBUF_ORDER, KPEX_RBUF_ORDER, m3::TCU::NO_REPLIES);
        });
        rbuf += KPEX_RBUF_SIZE;

        // configure upcall receive EP
        TCU::config_remote_ep(0, pe, m3::TCU::PEXUP_REP, [rbuf](m3::TCU::reg_t *ep_regs) {
            TCU::config_recv(ep_regs, m3::KIF::PEMUX_VPE_ID, rbuf,
                             PEXUP_RBUF_ORDER, PEXUP_RBUF_ORDER, m3::TCU::PEXUP_RPLEP);
        });
    }
    #endif
}

void PEMux::handle_call(const m3::TCU::Message *msg) {
    auto req = reinterpret_cast<const m3::KIF::PEXCalls::Exit*>(msg->data);
    capsel_t vpe = req->vpe_sel;
    int exitcode = req->code;

    KLOG(PEXC, "PEMux[" << peid() << "] got exit(vpe=" << vpe << ", code=" << exitcode << ")");

    auto vpecap = static_cast<VPECapability*>(_caps.get(vpe, Capability::VIRTPE));
    if(vpecap != nullptr) {
        vpecap->obj->_flags |= VPE::F_STOPPED;
        vpecap->obj->stop_app(exitcode, true);
    }

    // give credits back
    auto reply = m3::KIF::DefaultReply();
    reply.error = m3::Errors::NONE;
    TCU::reply(TCU::PEX_REP, &reply, sizeof(reply), msg);
}

void PEMux::add_vpe(VPECapability *vpe) {
    _caps.obtain(vpe->obj->id(), vpe);
    _vpes++;
}

void PEMux::remove_vpe(UNUSED VPE *vpe) {
    // has already been revoked
    assert(_caps.get(vpe->id(), Capability::VIRTPE) == nullptr);
    _vpes--;
    _mem_base = 0;
}

epid_t PEMux::find_eps(uint count) const {
    uint bit, start = _eps.first_clear();
    for(bit = start; bit < start + count && bit < EP_COUNT; ++bit) {
        if(_eps.is_set(bit))
            start = bit + 1;
    }
    if(bit != start + count)
        return EP_COUNT;
    return start;
}

bool PEMux::eps_free(epid_t start, uint count) const {
    for(epid_t ep = start; ep < start + count; ++ep) {
        if(_eps.is_set(ep))
            return false;
    }
    return true;
}

void PEMux::alloc_eps(epid_t first, uint count) {
    KLOG(EPS, "PEMux[" << peid() << "] allocating EPs " << first << " .. " << (first + count - 1));
    for(uint bit = first; bit < first + count; ++bit)
        _eps.set(bit, true);
}

void PEMux::free_eps(epid_t first, uint count) {
    KLOG(EPS, "PEMux[" << peid() << "] freeing EPs " << first << ".." << (first + count - 1));

    for(epid_t ep = first; ep < first + count; ++ep) {
        assert(_eps.is_set(ep));
        _eps.clear(ep);
    }
}

m3::Errors::Code PEMux::map(vpeid_t vpe, goff_t virt, m3::GlobAddr global, uint pages, uint perm) {
    m3::KIF::PEXUpcalls::Map req;
    req.opcode = static_cast<xfer_t>(m3::KIF::PEXUpcalls::MAP);
    req.vpe_sel = vpe;
    req.virt = virt;
    req.global = global.raw();
    req.pages = pages;
    req.perm = static_cast<xfer_t>(perm);

    KLOG(PEXC, "PEMux[" << peid() << "] sending map(vpe=" << req.vpe_sel
        << ", virt=" << m3::fmt((void*)req.virt, "p")
        << ", global=" << global << ", pages=" << req.pages << ", perm=" << req.perm << ")");

    return upcall(&req, sizeof(req));
}

m3::Errors::Code PEMux::translate(vpeid_t vpe, goff_t virt, uint perm, m3::GlobAddr *global) {
    m3::KIF::PEXUpcalls::Translate req;
    req.opcode = static_cast<xfer_t>(m3::KIF::PEXUpcalls::TRANSLATE);
    req.vpe_sel = vpe;
    req.virt = virt;
    req.perm = perm;

    KLOG(PEXC, "PEMux[" << peid() << "] sending translate(vpe=" << req.vpe_sel
        << ", virt=" << m3::fmt((void*)req.virt, "p") << ")");

    xfer_t val;
    m3::Errors::Code res = upcall(&req, sizeof(req), &val);
    if(res != m3::Errors::NONE)
        return res;
    *global = m3::GlobAddr(val & ~PAGE_MASK);
    return m3::Errors::NONE;
}

m3::Errors::Code PEMux::vpe_ctrl(VPE *vpe, m3::KIF::PEXUpcalls::VPEOp ctrl) {
    static const char *ctrls[] = {
        "INIT", "START", "STOP"
    };

    m3::KIF::PEXUpcalls::VPECtrl req;
    req.opcode = static_cast<xfer_t>(m3::KIF::PEXUpcalls::VPE_CTRL);
    req.vpe_sel = vpe->id();
    req.vpe_op = ctrl;
    req.eps_start = vpe->eps_start();

    KLOG(PEXC, "PEMux[" << peid() << "] sending VPECtrl(vpe="
        << req.vpe_sel << ", ctrl=" << ctrls[req.vpe_op] << ")");

    return upcall(&req, sizeof(req));
}

m3::Errors::Code PEMux::upcall(void *req, size_t size, xfer_t *val) {
    // send upcall
    event_t event = _upcqueue.send(0, req, size, false);
    m3::ThreadManager::get().wait_for(event);

    // wait for reply
    auto reply_msg = reinterpret_cast<const m3::TCU::Message*>(m3::ThreadManager::get().get_current_msg());
    auto reply = reinterpret_cast<const m3::KIF::PEXUpcalls::Response*>(reply_msg->data);
    if(val)
        *val = reply->val;
    return static_cast<m3::Errors::Code>(reply->error);
}

m3::Errors::Code PEMux::invalidate_ep(vpeid_t vpe, epid_t ep, bool force) {
    KLOG(EPS, "PE" << peid() << ":EP" << ep << " = invalid");

    uint32_t unread_mask;
    m3::Errors::Code res = TCU::inval_ep_remote(vpe, peid(), ep, force, &unread_mask);
    if(res != m3::Errors::NONE)
        return res;

    if(unread_mask != 0) {
        m3::KIF::PEXUpcalls::RemMsgs req;
        req.opcode = static_cast<xfer_t>(m3::KIF::PEXUpcalls::REM_MSGS);
        req.vpe_sel = vpe;
        req.unread_mask = unread_mask;
        return upcall(&req, sizeof(req));
    }
    else
        return m3::Errors::NONE;
}

void PEMux::notify_invalidate(vpeid_t vpe, epid_t ep) {
    m3::KIF::PEXUpcalls::EPInval req;
    req.opcode = static_cast<xfer_t>(m3::KIF::PEXUpcalls::EP_INVAL);
    req.vpe_sel = vpe;
    req.ep = ep;
    _upcqueue.send(0, &req, sizeof(req), false);
}

m3::Errors::Code PEMux::config_rcv_ep(epid_t ep, vpeid_t vpe, epid_t rpleps, RGateObject &obj) {
    assert(obj.activated());

    vpeid_t ep_vpe = Platform::is_shared(peid()) ? vpe : VPE::INVALID_ID;
    KLOG(EPS, "PE" << peid() << ":EP" << ep << " = "
        "RGate[vpe=" << ep_vpe << ", addr=#" << m3::fmt(obj.addr, "x")
        << ", order=" << obj.order
        << ", msgorder=" << obj.msgorder
        << ", replyeps=" << rpleps
        << "]");

    TCU::config_remote_ep(vpe, peid(), ep, [&obj, rpleps, ep_vpe](m3::TCU::reg_t *ep_regs) {
        TCU::config_recv(ep_regs, ep_vpe, obj.addr, obj.order, obj.msgorder, rpleps);
    });

    m3::ThreadManager::get().notify(reinterpret_cast<event_t>(&obj));
    return m3::Errors::NONE;
}

m3::Errors::Code PEMux::config_snd_ep(epid_t ep, vpeid_t vpe, SGateObject &obj) {
    assert(obj.rgate->addr != 0);
    if(obj.activated)
        return m3::Errors::EXISTS;

    vpeid_t ep_vpe = Platform::is_shared(peid()) ? vpe : VPE::INVALID_ID;
    KLOG(EPS, "PE" << peid() << ":EP" << ep << " = "
        "Send[vpe=" << ep_vpe << ", pe=" << obj.rgate->pe
        << ", ep=" << obj.rgate->ep
        << ", label=#" << m3::fmt(obj.label, "x")
        << ", msgsize=" << obj.rgate->msgorder
        << ", crd=#" << m3::fmt(obj.credits, "x")
        << "]");

    obj.activated = true;

    TCU::config_remote_ep(vpe, peid(), ep, [&obj, ep_vpe](m3::TCU::reg_t *ep_regs) {
        TCU::config_send(ep_regs, ep_vpe, obj.label, obj.rgate->pe, obj.rgate->ep,
                         obj.rgate->msgorder, obj.credits);
    });
    return m3::Errors::NONE;
}

m3::Errors::Code PEMux::config_mem_ep(epid_t ep, vpeid_t vpe, const MGateObject &obj, goff_t off) {
    if(off >= obj.size || obj.addr.raw() + off < off)
        return m3::Errors::INV_ARGS;

    vpeid_t ep_vpe = Platform::is_shared(peid()) ? vpe : VPE::INVALID_ID;
    KLOG(EPS, "PE" << peid() << ":EP" << ep << " = "
        "Mem [vpe=" << ep_vpe
        << ", addr=#" << (obj.addr + off)
        << ", size=#" << m3::fmt(obj.size - off, "x")
        << ", perms=#" << m3::fmt(obj.perms, "x")
        << "]");

    TCU::config_remote_ep(vpe, peid(), ep, [&obj, ep_vpe, off](m3::TCU::reg_t *ep_regs) {
        TCU::config_mem(ep_regs, ep_vpe, obj.addr.pe(),
                        obj.addr.offset() + off, obj.size - off, obj.perms);
    });
    return m3::Errors::NONE;
}

}

File: src/rustkernel/src/pes/pemux.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use base::col::{BitVec, Vec};
use base::errors::{Code, Error};
use base::goff;
use base::kif;
use base::mem::GlobAddr;
use base::rc::Rc;
use base::tcu::{self, EpId, PEId};

use cap::{MGateObject, PEObject, RGateObject, SGateObject};
use com::SendQueue;
use ktcu;
use pes::{VPEId, INVAL_ID};
use platform;

pub const MSG_ORD: u32 = 7;

pub struct PEMux {
    pe: Rc<PEObject>,
    vpes: Vec<VPEId>,
    queue: SendQueue,
    eps: BitVec,
    mem_base: goff,
}

impl PEMux {
    pub fn new(pe: PEId) -> Self {
        let mut pemux = PEMux {
            pe: PEObject::new(pe, (tcu::EP_COUNT - tcu::FIRST_USER_EP) as u32),
            vpes: Vec::new(),
            queue: SendQueue::new(pe as u64, pe),
            eps: BitVec::new(tcu::EP_COUNT),
            mem_base: 0,
        };

        for ep in 0..tcu::FIRST_USER_EP {
            pemux.eps.set(ep);
        }

        #[cfg(target_os = "none")]
        if platform::pe_desc(pe).supports_pemux() {
            pemux.init();
        }

        pemux
    }

    pub fn has_vpes(&self) -> bool {
        !self.vpes.is_empty()
    }

    pub fn add_vpe(&mut self, vpe: VPEId) {
        self.vpes.push(vpe);
    }

    pub fn rem_vpe(&mut self, vpe: VPEId) {
        assert!(self.vpes.len() > 0);
        self.vpes.retain(|id| *id != vpe);
    }

    #[cfg(target_os = "none")]
    fn init(&mut self) {
        use base::cfg;

        // configure send EP
        {
            ktcu::config_remote_ep(self.pe_id(), tcu::KPEX_SEP, |regs| {
                ktcu::config_send(
                    regs,
                    kif::pemux::VPE_ID as VPEId,
                    self.pe_id() as tcu::Label,
                    platform::kernel_pe(),
                    ktcu::KPEX_EP,
                    cfg::KPEX_RBUF_ORD,
                    1,
                );
            })
            .unwrap();
        }

        // configure receive EP
        let mut rbuf = platform::rbuf_pemux(self.pe_id());
        {
            ktcu::config_remote_ep(self.pe_id(), tcu::KPEX_REP, |regs| {
                ktcu::config_recv(
                    regs,
                    kif::pemux::VPE_ID as VPEId,
                    rbuf,
                    cfg::KPEX_RBUF_ORD,
                    cfg::KPEX_RBUF_ORD,
                    None,
                );
            })
            .unwrap();
            rbuf += 1 << cfg::KPEX_RBUF_ORD;
        }

        // configure upcall EP
        {
            ktcu::config_remote_ep(self.pe_id(), tcu::PEXUP_REP, |regs| {
                ktcu::config_recv(
                    regs,
                    kif::pemux::VPE_ID as VPEId,
                    rbuf,
                    cfg::PEXUP_RBUF_ORD,
                    cfg::PEXUP_RBUF_ORD,
                    Some(tcu::PEXUP_RPLEP),
                );
            })
            .unwrap();
        }
    }

    pub fn pe(&self) -> &Rc<PEObject> {
        &self.pe
    }

    pub fn pe_id(&self) -> PEId {
        self.pe.pe()
    }

    #[cfg(target_os = "linux")]
    pub fn eps_base(&mut self) -> goff {
        self.mem_base
    }

    pub fn set_mem_base(&mut self, addr: goff) {
        self.mem_base = addr;
    }

    pub fn find_eps(&self, count: u32) -> Result<tcu::EpId, Error> {
        let mut start = self.eps.first_clear();
        let mut bit = start;
        while bit < start + count as usize && bit < tcu::EP_COUNT {
            if self.eps.is_set(bit) {
                start = bit + 1;
            }
            bit += 1;
        }

        if bit != start + count as usize {
            Err(Error::new(Code::NoSpace))
        }
        else {
            Ok(start)
        }
    }

    pub fn eps_free(&self, start: tcu::EpId, count: u32) -> bool {
        for ep in start..start + count as usize {
            if self.eps.is_set(ep) {
                return false;
            }
        }
        true
    }

    pub fn alloc_eps(&mut self, start: tcu::EpId, count: u32) {
        klog!(
            EPS,
            "PEMux[{}] allocating EPS {}..{}",
            self.pe_id(),
            start,
            start as u32 + count - 1
        );
        for bit in start..start + count as usize {
            assert!(!self.eps.is_set(bit));
            self.eps.set(bit);
        }
    }

    pub fn free_eps(&mut self, start: tcu::EpId, count: u32) {
        klog!(
            EPS,
            "PEMux[{}] freeing EPS {}..{}",
            self.pe_id(),
            start,
            start as u32 + count - 1
        );
        for bit in start..start + count as usize {
            assert!(self.eps.is_set(bit));
            self.eps.clear(bit);
        }
    }

    fn ep_vpe_id(&self, vpe: VPEId) -> VPEId {
        match platform::is_shared(self.pe_id()) {
            true => vpe,
            false => INVAL_ID,
        }
    }

    pub fn config_snd_ep(
        &mut self,
        ep: EpId,
        vpe: VPEId,
        obj: &Rc<SGateObject>,
    ) -> Result<(), Error> {
        let rgate = obj.rgate();
        assert!(rgate.activated());

        klog!(EPS, "PE{}:EP{} = {:?}", self.pe_id(), ep, obj);

        ktcu::config_remote_ep(self.pe_id(), ep, |regs| {
            let vpe = self.ep_vpe_id(vpe);
            ktcu::config_send(
                regs,
                vpe,
                obj.label(),
                rgate.pe().unwrap(),
                rgate.ep().unwrap(),
                rgate.msg_order(),
                obj.credits(),
            );
        })
    }

    pub fn config_rcv_ep(
        &mut self,
        ep: EpId,
        vpe: VPEId,
        reply_eps: Option<EpId>,
        obj: &Rc<RGateObject>,
    ) -> Result<(), Error> {
        klog!(EPS, "PE{}:EP{} = {:?}", self.pe_id(), ep, obj);

        ktcu::config_remote_ep(self.pe_id(), ep, |regs| {
            let vpe = self.ep_vpe_id(vpe);
            ktcu::config_recv(
                regs,
                vpe,
                obj.addr(),
                obj.order(),
                obj.msg_order(),
                reply_eps,
            );
        })?;

        thread::ThreadManager::get().notify(obj.get_event(), None);
        Ok(())
    }

    pub fn config_mem_ep(
        &mut self,
        ep: EpId,
        vpe: VPEId,
        obj: &Rc<MGateObject>,
        pe_id: PEId,
        off: goff,
    ) -> Result<(), Error> {
        if off >= obj.size() as goff || obj.addr().raw().checked_add(off).is_none() {
            return Err(Error::new(Code::InvArgs));
        }

        klog!(EPS, "PE{}:EP{} = {:?}", self.pe_id(), ep, obj);

        ktcu::config_remote_ep(self.pe_id(), ep, |regs| {
            let vpe = self.ep_vpe_id(vpe);
            ktcu::config_mem(
                regs,
                vpe,
                pe_id,
                obj.offset() + off,
                (obj.size() - off) as usize,
                obj.perms(),
            );
        })
    }

    pub fn invalidate_ep(&mut self, ep: EpId, force: bool) -> Result<(), Error> {
        klog!(EPS, "PE{}:EP{} = invalid", self.pe_id(), ep);

        ktcu::invalidate_ep_remote(self.pe_id(), ep, force)
    }

    #[cfg(target_os = "none")]
    pub fn handle_call(&mut self, msg: &tcu::Message) {
        use pes::{vpemng, VPE};

        let req = msg.get_data::<kif::pemux::Exit>();
        let vpe_id = req.vpe_sel as VPEId;
        let exitcode = req.code as i32;

        klog!(
            PEXC,
            "PEMux[{}] got exit(vpe={}, code={})",
            self.pe_id(),
            vpe_id,
            exitcode
        );

        if self.vpes.contains(&vpe_id) {
            let vpe = vpemng::get().vpe(vpe_id).unwrap();
            VPE::stop_app(&vpe, exitcode, true);
        }

        let reply = kif::DefaultReply { error: 0 };
        ktcu::reply(ktcu::KPEX_EP, &reply, msg).unwrap();
    }

    #[cfg(target_os = "linux")]
    pub fn vpe_ctrl(
        &mut self,
        _vpe: VPEId,
        _eps_start: EpId,
        _ctrl: base::kif::pemux::VPEOp,
    ) -> Result<(), Error> {
        // nothing to do
        Ok(())
    }

    #[cfg(target_os = "none")]
    pub fn vpe_ctrl(
        &mut self,
        vpe: VPEId,
        eps_start: EpId,
        ctrl: base::kif::pemux::VPEOp,
    ) -> Result<(), Error> {
        let req = kif::pemux::VPECtrl {
            op: kif::pemux::Upcalls::VPE_CTRL.val as u64,
            vpe_sel: vpe as u64,
            vpe_op: ctrl.val as u64,
            eps_start: eps_start as u64,
        };

        klog!(
            PEXC,
            "PEMux[{}] sending VPECtrl(vpe={}, ctrl={:?})",
            self.pe_id(),
            vpe,
            ctrl
        );

        self.upcall(&req).map(|_| ())
    }

    #[cfg(target_os = "linux")]
    pub fn map(
        &mut self,
        _vpe: VPEId,
        _virt: goff,
        _glob: GlobAddr,
        _pages: usize,
        _perm: kif::PageFlags,
    ) -> Result<(), Error> {
        Ok(())
    }

    #[cfg(target_os = "none")]
    pub fn map(
        &mut self,
        vpe: VPEId,
        virt: goff,
        glob: GlobAddr,
        pages: usize,
        perm: kif::PageFlags,
    ) -> Result<(), Error> {
        let req = kif::pemux::Map {
            op: kif::pemux::Upcalls::MAP.val as u64,
            vpe_sel: vpe as u64,
            virt: virt as u64,
            global: glob.raw(),
            pages: pages as u64,
            perm: perm.bits() as u64,
        };

        klog!(
            PEXC,
            "PEMux[{}] sending Map(vpe={}, virt={:#x}, glob={:?}, pages={}, perm={:?})",
            self.pe_id(),
            vpe,
            virt,
            glob,
            pages,
            perm
        );

        self.upcall(&req).map(|_| ())
    }

    #[cfg(target_os = "none")]
    pub fn translate(
        &mut self,
        vpe: VPEId,
        virt: goff,
        perm: kif::Perm,
    ) -> Result<GlobAddr, Error> {
        use base::cfg::PAGE_MASK;

        let req = kif::pemux::Translate {
            op: kif::pemux::Upcalls::TRANSLATE.val as u64,
            vpe_sel: vpe as u64,
            virt: virt as u64,
            perm: perm.bits() as u64,
        };

        klog!(
            PEXC,
            "PEMux[{}] sending Translate(vpe={}, virt={:#x})",
            self.pe_id(),
            vpe,
            virt
        );

        self.upcall(&req)
            .map(|reply| GlobAddr::new(reply.val & !PAGE_MASK as goff))
    }

    #[cfg(target_os = "none")]
    fn upcall<R>(&mut self, req: &R) -> Result<&'static kif::pemux::Response, Error> {
        use base::util;

        let event = self
            .queue
            .send(tcu::PEXUP_REP, 0, util::object_to_bytes(req))?;
        thread::ThreadManager::get().wait_for(event);

        let reply = thread::ThreadManager::get().fetch_msg().unwrap();
        let reply = reply.get_data::<kif::pemux::Response>();
        if reply.error == 0 {
            Ok(reply)
        }
        else {
            Err(Error::new(Code::from(reply.error as u32)))
        }
    }

    #[cfg(target_os = "linux")]
    pub fn update_eps(&mut self) -> Result<(), Error> {
        ktcu::update_eps(self.pe_id(), self.mem_base)
    }
}
